docs/retries.md

```diff
@@ -0,0 +1,408 @@
+from pydantic_ai.providers.openai import OpenAIProvider
+
+def create_retrying_client():
+    """Create a client with smart retry handling for multiple error types."""
+
+    def should_retry_status(response):
+        """Raise exceptions for retryable HTTP status codes."""
+        if response.status_code in (429, 502, 503, 504):
+            response.raise_for_status()  # This will raise HTTPStatusError
+
+    transport = AsyncTenacityTransport(
+        controller=AsyncRetrying(
+            # Retry on HTTP errors and connection issues
+            retry=retry_if_exception_type((HTTPStatusError, ConnectionError)),
+            # Smart waiting: respects Retry-After headers, falls back to exponential backoff
+            wait=wait_retry_after(
+                fallback_strategy=wait_exponential(multiplier=1, max=60),
+                max_wait=300
+            ),
+            # Stop after 5 attempts
+            stop=stop_after_attempt(5),
+            # Re-raise the last exception if all retries fail
+            reraise=True
+        ),
+        validate_response=should_retry_status
+    )
+    return AsyncClient(transport=transport)
+
+# Use the retrying client with a model
+client = create_retrying_client()
+model = OpenAIModel('gpt-4o', provider=OpenAIProvider(http_client=client))
+agent = Agent(model)
+```
+
+## Wait Strategies
+
+### wait_retry_after
+
+The `wait_retry_after` function is a smart wait strategy that automatically respects HTTP `Retry-After` headers:
+
+```python {title="wait_strategy_example.py"}
+from pydantic_ai.tenacity import wait_retry_after
+from tenacity import wait_exponential
+
+# Basic usage - respects Retry-After headers, falls back to exponential backoff
+wait_strategy_1 = wait_retry_after()
+
+# Custom configuration
+wait_strategy_2 = wait_retry_after(
+    fallback_strategy=wait_exponential(multiplier=2, max=120),
+    max_wait=600  # Never wait more than 10 minutes
+)
+```
+
+This wait strategy:
+- Automatically parses `Retry-After` headers from HTTP 429 responses
+- Supports both seconds format (`"30"`) and HTTP date format (`"Wed, 21 Oct 2015 07:28:00 GMT"`)
+- Falls back to your chosen strategy when no header is present
+- Respects the `max_wait` limit to prevent excessive delays
+
+## Transport Classes
+
+### AsyncTenacityTransport
+
+For asynchronous HTTP clients (recommended for most use cases):
+
+```python {title="async_transport_example.py"}
+from httpx import AsyncClient
+from tenacity import AsyncRetrying, stop_after_attempt
+from pydantic_ai.tenacity import AsyncTenacityTransport
+
+# Create the basic components
+async_retrying = AsyncRetrying(stop=stop_after_attempt(3), reraise=True)
+
+def validator(response):
+    """Optional response validator."""
+    response.raise_for_status()
+
+# Create the transport
+transport = AsyncTenacityTransport(
+    controller=async_retrying,   # AsyncRetrying instance
+    validate_response=validator  # Optional response validator
```

Comment:
If this is not provided, under what circumstances does the retry logic actually come in?

Reply:
If not in place, it requires an exception to be raised. An example of something that would cause this is a request timeout. If you get a valid http response we need to actually trigger an error for a retry to happen (which is why these examples actually trigger it to raise if get a non-2XX status code).